### PR TITLE
Create signmessagewithprivkey rpc

### PR DESCRIPTION
**Pull request by achow101**
Since no one else did...

This adds a new RPC which does not rely on the wallet to sign messages. Continuation of #7899

### CODE REVIEW

File: src/rpc/misc.cpp [hunk 1]
```diff
@@ -366,6 +366,49 @@ UniValue verifymessage(const UniValue& params, bool fHelp)
     return (pubkey.GetID() == keyID);
 }
 
+UniValue signmessagewithprivkey(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() != 2)
+        throw runtime_error(
+            "signmessage \"privkey\" \"message\"\n"
+            "\nSign a message with the private key of an address"
+            + HelpRequiringPassphrase() + "\n"
```
Comment (paveljanik) [inline src/rpc/misc.cpp]: You do not need `HelpRequiringPassphrase` here. It is unavailable in no-wallet builds (see https://travis-ci.org/bitcoin/bitcoin/jobs/126001935#L1864).

File: src/rpc/misc.cpp [hunk 2]
```diff
@@ -366,6 +366,49 @@ UniValue verifymessage(const UniValue& params, bool fHelp)
     return (pubkey.GetID() == keyID);
 }
 
+UniValue signmessagewithprivkey(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() != 2)
+        throw runtime_error(
+            "signmessage \"privkey\" \"message\"\n"
```
Comment (paveljanik) [inline src/rpc/misc.cpp]: This has to be `signmessagewithprivkey` instead of `signmessage`.

File: src/rpc/misc.cpp (hunk 1 above)
Comment (achow101) [inline src/rpc/misc.cpp]: Fixed

File: src/rpc/misc.cpp (hunk 2 above)
Comment (achow101) [inline src/rpc/misc.cpp]: Fixed

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/7953/commits/f90efbfeef9cd733aa7590fdf581ff82b999c4b5

File: src/rpc/misc.cpp [hunk 3]
```diff
@@ -366,6 +366,48 @@ UniValue verifymessage(const UniValue& params, bool fHelp)
...
+    CHashWriter ss(SER_GETHASH, 0);
+    ss << strMessageMagic;
+    ss << strMessage;
+
+    vector<unsigned char> vchSig;
+    if (!key.SignCompact(ss.GetHash(), vchSig))
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Sign failed");
+
+    return EncodeBase64(&vchSig[0], vchSig.size());
+}
```
Comment (jonasschnelli) [inline src/rpc/misc.cpp]: IMO `strPrivkey`, `vchSecret` and `params[0]` should be mem-cleansed before the function returns.
I guess `key` RAII -cleans itself.

**Comment by jonasschnelli** [comment]
utACK f90efbfeef9cd733aa7590fdf581ff82b999c4b5

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/commit/f90efbfeef9cd733aa7590fdf581ff82b999c4b5

@jonasschnelli Right. We should also recommend using `-stdin` for this when document this new RPC in release notes.

File: src/rpc/misc.cpp (hunk 3 above)
Comment (laanwj) [inline src/rpc/misc.cpp]: - vchSecret is ok - CBitcoinSecret (derived from CBase58Data) uses a zero_after_free allocator
- strPrivKey could be a SecureString, I guess.

However, params[] is a const parameter vector passed in, provided by a third-party library. I don't think this issue can be solved in general. The http server, JSON parsing, the RPC mechanism leaves the parameters all over memory. We have a similar problem in importprivkey. No need to solve in this pull.

Comment (jonasschnelli) [inline src/rpc/misc.cpp]: Agree. Its just a nitpick (I forgot to pre-fix it with "nit:"). As long as we are dealing with privatekeys in bitcoin-core there will always be memory containing sensitive stuff.

**Comment by paveljanik** [comment]
Can you please write some test for this new RPC?

**Comment by achow101** [comment]
@paveljanik I've added a test. I think I did it correctly.

File: qa/rpc-tests/signmessages.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    def run_test(self):
+        message = 'This is just a test message'
+
+        # Test the signing with a privkey
+        privKey = 'cUeKHd5orzT3mz8P9pxyREHfsWtVfgsfDjiZZBcjUBAaGk1BTj7N'
+        address = 'mpLQjfK79b7CCV4VMJWEWAj5Mpx8Up5zxB'
+        try:
+            signature = self.nodes[0].signmessagewithprivkey(privKey, message)
+        except JSONRPCException as e:
+            assert("Invalid private key" not in e.error['message'])
```
Comment (MarcoFalke) [inline qa/rpc-tests/signmessages.py]: All of those seem dead code, so you might as well remove them?

Comment (MarcoFalke) [inline qa/rpc-tests/signmessages.py]: What I mean is: If you don't expect an exception, don't catch it. The whole purpose of the test framework is to fail on any unexpected exception.

Other than that, you may want to try this syntax:

```py
try:
  something_that_fails()
  assert(False)
except:
  assert(thing_on_failure())
```

or

```py
try:
  something_that_fails()
except:
  assert(thing_on_failure())
else:
  assert(False)

Comment (achow101) [inline qa/rpc-tests/signmessages.py]: Ok, I'll remove them. I should probably add a failure condition, right?

File: qa/rpc-tests/signmessages.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+        # Verify the message
+        self.nodes[0].verifymessage(address, signature, message)
+
+        # Test the signing with an address with wallet
+        address = self.nodes[0].getnewaddress()
+        signature = self.nodes[0].signmessage(address, message)
+
+        # Verify the message
+        self.nodes[0].verifymessage(address, signature, message)
```
Comment (laanwj) [inline qa/rpc-tests/signmessages.py]: You should assert the result from verifymessage

File: qa/rpc-tests/signmessages.py (hunk 4 above)
Comment (MarcoFalke) [inline qa/rpc-tests/signmessages.py]: Right, if you feel like adding a case that should fail to check if it fails properly, you are very welcome to do so.

File: qa/rpc-tests/signmessages.py (hunk 5 above)
Comment (achow101) [inline qa/rpc-tests/signmessages.py]: done

**Comment by jonasschnelli** [comment]
Tested ACK.
Binaries: https://bitcoin.jonasschnelli.ch/pulls/7953/

**Pull request merged**